### PR TITLE
Implement Ancestor Score Index in Mempool

### PR DESCRIPTION
**Pull request by royster57**
This will be used for selecting the most economically attractive entries for block production.

### CODE REVIEW

File: mempool/src/pool.rs [hunk 1]
```diff
@@ -385,10 +377,18 @@ where
             .filter_map(|input| input.outpoint().tx_id().get_tx_id().cloned())
             .filter_map(|id| self.store.txs_by_id.contains_key(&id).then_some(id))
             .collect::<BTreeSet<_>>();
+        let ancestor_ids =
+            TxMempoolEntry::unconfirmed_ancestors_from_parents(parents.clone(), &self.store)?;
+        let ancestors = ancestor_ids
+            .0
```
Comment (stanislav-tkach) [inline mempool/src/pool.rs:383]: It isn't related to your pull request, but I find this "index field access" somewhat confusing: I immediately want to check why it is a tuple and why only one field is utilized. Our `newtype` macro generates the `Deref` implementation, but `*ancestor_ids.into_iter()` breakes the type inference. Maybe we want to add a method such as `inner` to the `newtype` macro or is it only me who is frustrated in such cases?

File: mempool/src/pool/store.rs [hunk 2]
```diff
@@ -77,10 +85,21 @@ pub struct MempoolStore {
     pub(super) spender_txs: BTreeMap<OutPoint, Id<Transaction>>,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(super) enum MempoolRemovalReason {
+    Block,
+    //Conflict,
+    Expiry,
+    SizeLimit,
+    //Reorg,
```
Comment (stanislav-tkach) [inline mempool/src/pool/store.rs:94]: Should the commented lines be removed?

**Review by stanislav-tkach — APPROVED** [review]

File: mempool/src/pool/store.rs [hunk 3]
```diff
@@ -64,6 +69,9 @@ pub struct MempoolStore {
     // rather a value that takes into account witdess and sigop data (see CTxMemPoolEntry::GetTxSize).
     pub(super) txs_by_descendant_score: BTreeMap<DescendantScore, BTreeSet<Id<Transaction>>>,
 
+    // Mempool entries sorted by ancestor score
```
Comment (TheQuantumPhysicist) [inline mempool/src/pool/store.rs:72]: Maybe it's a good place to explain what ancestor score means

File: mempool/src/pool.rs (hunk 1 above)
Comment (royster57) [reply mempool/src/pool.rs:383]: Opened an issue for this.

File: mempool/src/pool/store.rs [hunk 4]
```diff
@@ -77,10 +85,21 @@ pub struct MempoolStore {
     pub(super) spender_txs: BTreeMap<OutPoint, Id<Transaction>>,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(super) enum MempoolRemovalReason {
```
Comment (TheQuantumPhysicist) [inline mempool/src/pool/store.rs:89]: I'm confused... why does this even exist?

File: mempool/src/pool/store.rs (hunk 2 above)
Comment (royster57) [reply mempool/src/pool/store.rs:94]: We'll eventually use these as well, but I guess we can delete them for now.

File: mempool/src/pool/store.rs (hunk 3 above)
Comment (TheQuantumPhysicist) [reply mempool/src/pool/store.rs:72]: To rephrase, who cares why transactions are removed?

Comment (royster57) [reply mempool/src/pool/store.rs]: Updated the documentation.

File: mempool/src/pool.rs (hunk 1 above)
Comment (stanislav-tkach) [reply mempool/src/pool.rs:383]: It is just my personal opinion, so if everyone is OK with that, then I can live with that too. 🙃

File: mempool/src/pool/store.rs (hunk 4 above)
Comment (TheQuantumPhysicist) [reply mempool/src/pool/store.rs:89]: If you wanna keep this, please provide full documentation on it why it exists and what every arm does

File: mempool/src/pool/store.rs [hunk 5]
```diff
@@ -55,15 +60,18 @@ pub struct MempoolStore {
...
-    //
-    // TODO: currently, the descendant score is the sum fee of the transaction to gether with all
-    // of its descendants. If we wish to follow Bitcoin Core, we should use:
-    // max(feerate(tx, tx_with_descendants)),
-    // Where feerate is computed as fee(tx)/size(tx)
-    // Note that if we wish to follow Bitcoin Bore, "size" is not simply the encoded size, but
+    // The descendant score of an entry is defined as:
+    //  max(fee/size of entry's tx, fee/size with all descendants).
+    //  TODO if we wish to follow Bitcoin Bore, "size" is not simply the encoded size, but
     // rather a value that takes into account witdess and sigop data (see CTxMemPoolEntry::GetTxSize).
```
Comment (TheQuantumPhysicist) [inline mempool/src/pool/store.rs:66]: witness

File: mempool/src/pool/store.rs (hunk 4 above)
Comment (royster57) [reply mempool/src/pool/store.rs:89]: I copied this enum from Bitcoin Core. They have a mechanism for notifying clients interested in knowing when a transaction was removed for a reason other than inclusion in a block. Also, the Bitcoin Core wallet has some logic for transactions removed from the mempool due to conflicting transactions in a new block.
If this enum is confusing then I think there's no need for it at the moment. Replaced it with a boolean `remove_for_block`, which is all we really need in our logic.

File: mempool/src/pool/store.rs (hunk 5 above)
Comment (royster57) [reply mempool/src/pool/store.rs]: Fixed.

File: mempool/src/pool/store.rs (hunk 4 above)
Comment (TheQuantumPhysicist) [reply mempool/src/pool/store.rs]: Booleans are usually a downgrade. Just some documentation helps.

Comment (royster57) [reply mempool/src/pool/store.rs:96]: Ok, restored it and added some docs.

**Pull request merged**